codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/AwsJson10ProtocolGenerator.java

```diff
@@ -86,10 +97,156 @@ private void generateSharedDeserializers(GenerationContext context, GoWriter wri
+
+        generateOperationErrorDeserializers(context, writer, ops);
+
+        writer.write(getProtocolErrorInfo());
+        writer.write(GET_AWS_QUERY_ERROR_CODE);
+    }
+
+    private Set<Shape> generateErrorDeserializers(GenerationContext context, GoWriter writer, Set<OperationShape> ops) {
+        Set<Shape> errorShapes = new HashSet<>();
+        for (var op : ops) {
+            var errors = op.getErrors();
+            for (var error : errors) { //consider doing this directly in deserializeOperationErrors
+                Set<Shape> shapes = getShapesToSerde(context.getModel(), context.getModel().expectShape(error));
+                errorShapes.addAll(shapes);
+            }
+        }
+        return errorShapes;
+    }
+
+    private void generateOperationErrorDeserializers(
+            GenerationContext context, GoWriter writer, Set<OperationShape> operations) {
+        for (var operation : operations) {
+            writer.write(goTemplate("""
+                func $func:L(resp $smithyhttpResponse:P) error {
+                    payload, err := $readAll:T(resp.Body)
+                    if err != nil {
+                        return &$deserError:T{Err: $fmtErrorf:T("read response body: %w", err)}
+                    }
+
+                    typ, msg, v, err := getProtocolErrorInfo(payload)
+                    if err != nil {
+                        return &$deserError:T{Err: $fmtErrorf:T("get error info: %w", err)}
+                    }
+
+                    if len(typ) == 0 {
+                        typ = "UnknownError"
+                    }
+                    if len(msg) == 0 {
+                        msg = "UnknownError"
+                    }
+
+                    _ = v
+                    switch string(typ) {
+                    $errors:W
+                    default:
+                        $awsQueryCompatible:W
+                        return &$genericAPIError:T{Code: typ, Message: msg}
+                    }
+                }
+                """,
+                    MapUtils.of(
+                            "deserError", SmithyGoDependency.SMITHY.pointableSymbol("DeserializationError"),
+                            "fmtErrorf", GoStdlibTypes.Fmt.Errorf,
+                            "func", ProtocolGenerator.getOperationErrorDeserFunctionName(operation,
+                                    context.getService(), "awsJson10"),
+                            "genericAPIError", SmithyGoDependency.SMITHY.pointableSymbol("GenericAPIError"),
+                            "readAll", SmithyGoDependency.IO.func("ReadAll"),
+                            "smithyhttpResponse", SmithyGoTypes.Transport.Http.Response,
+                            "awsQueryCompatible", context.getService().hasTrait(AwsQueryCompatibleTrait.class)
+                                    ? deserializeAwsQueryError()
+                                    : emptyGoTemplate(),
+                            "errors", GoWriter.ChainWritable.of(
```

Comment:
Add this as a variable outside of the map so it's easier to follow, something like this

```java
for(var operation: operations) {
    var errors = context.getService()
        .stream()
        .map(it -> deserializeErrorCase(context, context.getModel().expectShape(it, StructureShape.class)))
        .toList()
}
```

Reply:
good idea